lib/zclient.c

```diff
@@ -967,6 +967,51 @@ int zapi_nexthop_encode(struct stream *s, const struct zapi_nexthop *api_nh,
 	return ret;
 }
 
+extern void zclient_nhg_del(struct zclient *zclient, uint32_t id)
+{
+	struct stream *s = zclient->obuf;
+
+	stream_reset(s);
+	zclient_create_header(s, ZEBRA_NHG_DEL, VRF_DEFAULT);
+
+	stream_putw(s, zclient->redist_default);
+	stream_putl(s, id);
+
+	stream_putw_at(s, 0, stream_get_endp(s));
+}
+
+static void zclient_nhg_writer(struct stream *s, uint16_t proto, int cmd,
+			       uint32_t id, size_t nhops,
+			       struct zapi_nexthop *znh)
```

Comment:
Hmm - can you say something about why this doesn't follow the common pattern of using a public zapi struct that's used for marshalling and un-ing? There are some benefits to that - a single place to put key validation logic, like "nhops must be < MULTIPATH". without that, it seems like every daemon who wants to work with nhgs will have to have its own copy of validation and error-handling logic. Those structs are also sort of self-documenting - if we need to add something to the zapi nhg, there'd be one place that defined that - for example, when backup nexthops need to be added.

Reply:
So im not sure I follow this comment? I think we are doing it pretty similarly to how its done for routes? Also, its worth noting that each daemons notion of a `nexthop/nexthop group` is very different between protos. For example, sharpd and bgp (the only users of this API for right now) dont share common structs that could call into this code. The way sharpd turns something into a nexthop group is different from how bgp does it.